### PR TITLE
Truetype Font Hinting

### PR DESCRIPTION
**Pull request by axkibe**
Implementing a font hinting interpreter for truetype:

Demo:

The actual canvas is the top one, the lower one has each pixel enlarged by a factor of 7.
The font used in the demo is DejaVuSans.ttf
First line is default opentype.js
Second line is opentype.js with fonthinting enabled
Third line is canvas.fillText() with @font-face.

hintingdemo

As you see there is hardly a difference between second and third line, except glyph advancement not being hinted. I didn't do that yet, as that would require some changes to the opentype.js.

I tried to keep all changes to opentype.js outside of src/hintingtt.js as minimal as possible. 

PS: I'll need to append a fix commits until the code styling checks stop complaining.

### CODE REVIEW

**Comment by axkibe** [comment]
Finally a green check :)

File: src/glyph.js [hunk 1]
```diff
@@ -118,19 +119,41 @@ Glyph.prototype.getBoundingBox = function() {
...
+        hPoints = font.hinting.exec(this, fontSize);
+    }
+
+    if (hPoints) {
+        // in case the hinting engine failed revert to
+        // plain rending
+        commands = glyf.getPath(hPoints).commands;
+        x = Math.round(x);
+        y = Math.round(y);
+        // TODO in case of hinting xyScaling is not yet supported
```
Comment (Jolg42) [inline src/glyph.js]: I don't understand this TODO, could you explain it a little bit more?

File: src/hintingtt.js [hunk 2]
```diff
@@ -0,0 +1,3043 @@
+/* A TrueType font hinting interpreter.
+*
+* (c) 2017 Axel Kittenberger
+*
+* This file is covered by the MIT (expat) License,
+* see the LICENSE file in the root of this project.
```
Comment (Jolg42) [inline src/hintingtt.js]: I don't if it's relevant to add the mention to the MIT license as the LICENSE file is here for that

File: index.html [hunk 3]
```diff
@@ -115,7 +115,7 @@
 }
 
 function fontSizeChanged() {
-    fontSize = fontSizeSlider.value;
+    fontSize = parseInt(fontSizeSlider.value, 10);
```
Comment (Jolg42) [inline index.html]: Just one question, is it doable to replace `parseInt` by `parseFloat`?

**Comment by Jolg42** [comment]
Your screenshot comparing the hinting looks great! :+1: 

I see that hinting is not activated by default, `Font.prototype.getPath` needs `options.hinting` to be true, I think it's the best to not impact memory consumption & performance for people that don't need it!

By the way, you don't have to push commits to check code styling, it's a lot easier to do `npm run test` (you can look at scripts, in [package.json](https://github.com/nodebox/opentype.js/blob/master/package.json))

->You worked a lot! Thank you for that contribution! :smiley:

File: src/glyph.js (hunk 1 above)
Comment (axkibe) [inline src/glyph.js]: The options object can have an xScale or yScale setting to stretch the paths, if hinting is enabled it is currently ignored.

File: src/hintingtt.js (hunk 2 above)
Comment (axkibe) [inline src/hintingtt.js]: I agree it's not important. You can cut it if you want to. It's oft generally recommended a good idea for every file tough. src/opentype.js has for example.

File: index.html (hunk 3 above)
Comment (axkibe) [inline index.html]: Sure! I'll append it.

**Comment by fdb** [comment]
Thanks, this looks awesome!

**Comment by axkibe** [comment]
Exactly, thats why I made it off by hinting default. It's only relevant between aprox. 8ppem and 60ppem, below it's a blur anyway and above normal subpixels are beautiful enough. Oh and I noted some font programs turn themselves off.

PS: This all applies only to truetype, postscript fonts have a completely different hinting engine.

**Comment by axkibe** [comment]
PPS: DejaVuSans has one of the best font hinting programs I encountered, other fonts the difference may not be that staggering.

**Comment by Jolg42** [comment]
As it is now merged, next step could be editing the README & add a short description of current hinting with an example :)

**Comment by fdb** [comment]
Good point! @axkibe could you provide an example for the README?

**Comment by axkibe** [comment]
OK.

**Comment by Jolg42** [comment]
@fdb The git history is a little verbose now, maybe you can try to squash next time @fdb :wink:

**Comment by fdb** [comment]
Oh right the code styling commits. Too late now, thanks for reminding me :blush:

**Comment by axkibe** [comment]
Im sorry.

**Comment by fdb** [comment]
@axkibe no problem, don't worry about it 😄

File: index.html (hunk 3 above)
Comment (Jolg42) [inline index.html]: Sorry for that, I checked and as the slider is configured to only output integers parseFloat is not needed...!

File: src/glyph.js (hunk 1 above)
Comment (Jolg42) [inline src/glyph.js]: Sure!

**Comment by Jolg42** [comment]
@fdb @axkibe I added the hinting checkbox for http://opentype.js.org/ to try it ;)

**Comment by axkibe** [comment]
Cool, unfortunially the default FireSansMedium.woff is a postscript font, so it works when loading a ttf font. But then it does :)

**Comment by axkibe** [comment]
I noticed DEBUG is turned on the live website.
This seriously dampens perfomance.

**Comment by Jolg42** [comment]
@axkibe  Fixed :) I forgot to rebuild the dist files....

**Comment by Jolg42** [comment]
@axkibe so maybe we could disable the hinting checkbox for non ttf fonts

**Pull request merged**